### ISSUE TITLE
RFC: Declare Maps Up-Front

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#2954]
This is a proposal to introduce a new syntax for declaring maps ahead of their use, outside of probes:
```
let @x[uint64] = stack_t;

BEGIN
{
  @x[123] = ustack();
}
```

I would like this functionality, but am not tied to this syntax in particular.

### Use Case 1 - Declare key/value types

The types of map keys and values are currently inferred automatically from usage. This works well for simple scripts, but there are maintainability benefits to declaring the expected types of variables in larger scripts, both for understandability and ease of refactoring.

#### Example - Types not inferred from first usage
```
let @x[uint64] = int64;

PROBE1
{
  @x["123"] = 456; // with a map declaration this line produces a type error
}
PROBE2
{
  @x[123] = 456;
}
PROBE3
{
  @x[0] = 0;
}
```
In the middle of refactoring a large script, changing the type of the key in PROBE1 might have been missed and it would be useful to get an error on that line. Without a map declaration, PROBE2 and PROBE3 would trigger type errors as PROBE1 was seen first.

#### Example - Control exact types used
```
let @x[uint8] = int32;

BEGIN
{
  @x[123] = 456; // the literals would be inferred as int64 without the declaration
}
```

### Use Case 2 - Configure Map Settings

The default capacity of a hash map is not large enough for many purposes. Increasing a map's capacity is currently only possible through a global config option that applies to all maps. This wastes memory by increasing the capacity of all maps, when other maps may be fine with a smaller capacity.

#### Example - Custom size for hash map
```
let @x[int64] = int64 : HashMap(1024);

BEGIN
{
  @x[123] = 456;
}
```

### Use Case 3 - Create non-hash-maps

Allow users to declare maps backed by data structures other than a BPF Hash Map. For example, an array map might be used for efficiency when it is known that a fixed selection of keys is needed.

#### Example - Create an array map
```
let @x[uint32] = int64 : ArrayMap(128); // BPF specifies that array maps must have a "uint32" key

BEGIN
{
  @x[123] = 456;
}
```

**Comment by danobi** [comment]
Yeah, use cases seem totally reasonable. Exact syntax should hopefully leverage type annotation work done for user defined functions.

**Comment by viktormalik** [comment]
Agreed with @danobi, #2624 uses the `$arg: type` syntax for function arguments so we should somehow mirror that here (we could eventually add `$var: type = ...` syntax for local vars.

**Comment by jordalgo** [comment]
All makes sense to me. Maybe because I just worked on the config syntax, I'm considering the "floating" element of these type annotations e.g. will they be forced to be declared at the top? should they go after `config = {`?

Perhaps they should get wrapped in some other similar demarcation e.g.
```
types = {
   let @x[uint64] = int64;
}
```
User-defined command line args could also live here. Just something to think about as bpftrace has no (and doesn't want) header files and we want a way to keep things organized.

**Comment by JakeHillion** [comment]
Taking a look at this now. The current (verbose) syntax I'm looking at is:

```
let global @x: int32;
let global @x: [int32 -> int64];
let global @x: [int32, int16 -> int64];
```

This would be extensible in future to (with certain conditions):
```
let global @y: int64 = ArrayMap;
let global @y = ArrayMap;
```

I considered making the global optional, but for now have left it global.

Placement rules are currently equivalent to probes. That is, you can place these let statements anywhere in the global scope. The semantic analyser rejects duplicate lets for the same map, and processes all maps before processing the rest of the program. This means that a map declared after a probe that uses it has full effect.

The let syntax could be further extensible to things like:
```
let t1 = [int32 -> int64];
let global @x: type;
let t2 = int32;
BEGIN { let y: t2; }
```

Feedback seems to be to drop the global, which doesn't create any more ambiguity so that's fine. I like the rest for a uniform "declare" something syntax but could be swayed too. Will put up a PR soon.

**Comment by jordalgo** [comment]
@JakeHillion Thanks for picking this up.

As discussed offline, I'm curious if dropping both the "global" and the "let" is possible to improve the conciseness of this new syntax e.g.

```
@x: int32;
@x: [(int32, int16) -> int64];
```

I think this could then be re-used for scratch variables inside probes when we want to add variable declarations e.g.

```
kprobe:do_nanosleep {
  $scratch: macaddr_t;
  if (condition) {
    $scratch = ...
  } else {
    $scratch = ...
  }
}
```

**Comment by jordalgo** [comment]
Restarting this discussion on map declarations as it seems to be of a higher priority now.

After the decision to add the `let` syntax: https://github.com/bpftrace/bpftrace/issues/3423
I propose this for basic map declarations:

Scalar maps (maps with no keys)
```
let @x: uint64; // no initialization
let @x: uint64 = 1; // with initialization
```

Maps with keys:
```
let @y: <string, uint64>; // no initialization
let @y: <string, uint64> = {{"hello", 1}, {"bye", 2}}; // with initialization
```
Don't want to use brackets here for the map type as those are reserved for arrays. Also as a first step we don't have to allow initialization with declaration for maps with keys.

Maps with multi-value keys:
```
let @z: <(string, int8), uint64>;
```

Something to note here is that, unlike scratch variables, declaring maps without attributes is somewhat useless. Unless we want to allow pre-populating a map with values (like in the example above) but you don't need a declaration for that - could just be some added sugar. The `let` keyword is also a tad suspect as maps don't suffer from block scoping so there is no case when a user would need to declare a map like `let @y;`. 

As per some discussion here: https://github.com/bpftrace/bpftrace/issues/3360) we would add map attributes so users can define what kind of BPF map type they want to use e.g.
```
[type = LRUHashMap]
let @z: <(string, int8), uint64>;
```

**Comment by ajor** [comment]
> Scalar maps (maps with no keys)
> ```
> let @x: uint64; // no initialization
> let @x: uint64 = 1; // with initialization
> ```

Looks good to me. Initialisation could be a fairly big extension but it seems reasonable to allow it.

> ```
> [type = LRUHashMap]
> let @z: <(string, int8), uint64>;
> ```

I'm thinking that the type of the map should appear directly in the type declaration rather than as an attribute - it is the "type" after all! Maybe like:
```
let @map: LRUHashMap<string, uint64>;
```
The syntax is starting to look pretty familiar 😅

Then my question is do we leave `max_entries` as an optional attribute for all map types, or should an array map require it to be specified? e.g.:
```
let @map: Array<int32, 32>; // 32 elements long
```

**Comment by danobi** [comment]
> > Scalar maps (maps with no keys)
> > ```
> > let @x: uint64; // no initialization
> > let @x: uint64 = 1; // with initialization
> > ```
> 
> Looks good to me. Initialisation could be a fairly big extension but it seems reasonable to allow it.
> 
> > ```
> > [type = LRUHashMap]
> > let @z: <(string, int8), uint64>;
> > ```
> 
> I'm thinking that the type of the map should appear directly in the type declaration rather than as an attribute - it is the "type" after all! Maybe like:
> 
> ```
> let @map: LRUHashMap<string, uint64>;
> ```

I think this is the important question here. B/c if we decide this is the best path forward, then we need to take on map key/value type declarations along with underlying map type in the same go.

But if we decide the two are decoupled (ie. with a map attribute), then we can take on the two things independently. In which case I'd strongly believe we should attack underlying map type first. But that is contingent on what we think the long term solution is.

**Comment by viktormalik** [comment]
> But if we decide the two are decoupled (ie. with a map attribute), then we can take on the two things independently. In which case I'd strongly believe we should attack underlying map type first. But that is contingent on what we think the long term solution is.

I don't have many concrete use-cases in mind right now so I may be missing some context but I am (as always) calling out for bpftrace simplifying user's lives as much as possible. That is, if the user wants to use a particular map type, we still shouldn't force him to specify full map signature including keys and value types, max entries, etc. So, from my perspective, these should be independent.

A counter-argument would be that if the user is expert enough to know which kind of map they want, they may as well fill the other information in. I sort of agree with that but still would prefer bpftrace to infer things where possible.

> Then my question is do we leave `max_entries` as an optional attribute for all map types, or should an array map require it to be specified? e.g.:
> 
> ```
> let @map: Array<int32, 32>; // 32 elements long
> ```

Other maps have `max_entries`, too. We may then end up with something like:
```
let @map: PercpuHashMap<int64, int64, 32>
```
which starts to be a bit unintuitive and named args would help a lot here.

OTOH I agree with @ajor that map type (as in "hash", "array", etc.) is a part of the overall map type (as in "data type of the storage") and therefore shouldn't be set from a different syntax construct. What about allowing optional argument names in the angle brackets? All these would be valid, then:
```
let @map: PercpuHashMap<key=int64, value=int64, max_entries=32>;
let @map: PercpuHashMap<int64, int64, 32>;
let @map: <key=int64, value=int64, max_entries=32>;
let @map: <int64, int64>;
```
Some of them are a bit long but I believe terseness is no longer a point for this syntax. We could also simplify our lives a bit by requiring either all or none of the args to have names but that mostly depends on how hard this would be to parse.

**Comment by jordalgo** [comment]
I like the addition of the map "type" in front of the angle bracket but I agree with @viktormalik that it would be nice if users didn't have to specify the key/value types in order to set the map "type" so maybe this is ok syntax:
```
let @map: PercpuHashMap;
// or 
let @map: PercpuHashMap<>;
```

This would potentially allow us to decouple the implementation but, at the same time, adding optional key/value types doesn't seem like it would be that much work or code.

---

> `let @map: PercpuHashMap<key=int64, value=int64, max_entries=32>;`

Do we feel like including `max_entries` and other map attributes within the angle brackets scales? Especially with named arguments that declaration might start to get long and ugly. Whereas if we go with putting attributes above the declaration then we can have line breaks e.g.
```
[max_entries = 32]
[other_attr = "whatever"]
let @z: PercpuHashMap<int8, uint64>;
```

Also if we go with probe attributes at some point there will be continuity there with map attributes, which I like.

**Comment by viktormalik** [comment]
> 
> > `let @map: PercpuHashMap<key=int64, value=int64, max_entries=32>;`
> 
> Do we feel like including `max_entries` and other map attributes within the angle brackets scales? Especially with named arguments that declaration might start to get long and ugly. Whereas if we go with putting attributes above the declaration then we can have line breaks e.g.
> 
> ```
> [max_entries = 32]
> [other_attr = "whatever"]
> let @z: PercpuHashMap<int8, uint64>;
> ```
> 
> Also if we go with probe attributes at some point there will be continuity there with map attributes, which I like.

Where do we draw the line between what's part of a map type and what's an "attribute"? In kernel, all the information (usually `type`, `key`, `value`, `max_entries` but some map types have other fields) is a part of the map type definition. Do we say that in bpftrace, just `key`, `value`, and `type` are a part of the type (i.e. in angle brackets) and everything else is an attribute? What if there are maps which require the other fields to be set and there's no "default" value?

**Comment by ajor** [comment]
RE `max_entries`: For hash maps, stacks, queues, etc., users generally won't care what the exact capacity is as long as big enough to hold the entries they want to insert. It's fine for the capacity to be bigger than needed.

For arrays, there's no difference between capacity and current size, so it is probably something that users should always set themselves. The capacity being bigger than needed will impact on user programs (other than memory usage): `len(@myarray)` will always be `max_entries` and for-loops will always loop that number of times.

That's why I think it'd be good to make `max_entries` a required field for arrays but keep it optional for other map types. This could be made to work with either syntax option though.

**Comment by danobi** [comment]
There's also flags to consider: https://github.com/torvalds/linux/blob/b1b46751671be5a426982f037a47ae05f37ff80b/tools/include/uapi/linux/bpf.h#L1357-L1412

I suspect it's unavoidable in the long run that kernel will add all sorts of fancy flags / fields for special maps. So having a good way to compose additional attributes is a must.

**Comment by JakeHillion** [comment]
This has all been going into a sidebox, my bad! I was going to dig out my branch, but it seems the syntax has changed enough that it's not worth it - most of the change was in the parser, actually implementing it (without specifying a map implementation) was easy, and I still hadn't done the remaining hard bit of getting decent error messages.

> Do we feel like including `max_entries` and other map attributes within the angle brackets scales? Especially with named arguments that declaration might start to get long and ugly. Whereas if we go with putting attributes above the declaration then we can have line breaks e.g.
> 
> ```
> [max_entries = 32]
> [other_attr = "whatever"]
> let @z: PercpuHashMap<int8, uint64>;
> ```
> 
> Also if we go with probe attributes at some point there will be continuity there with map attributes, which I like.

@jordalgo is there any reason we can't have the line breaks like:
```
let @z : PercpuHashMap<int8, uint64,
    max_entries = 64,
    other_attr = "whatever",
>;
```

Should all parse as whitespace right?

As an aside, and not to jump into another syntax debate, I think that's my preferred syntax:
- Starts with `let @z :` - familiar if you've seen variable decls in a lot of languages, you're declaring something named `@z`. `@z: ` by itself takes me a lot of mental parsing and I'd need to read the rest of the statement to work out I was declaring something variable like.
- Immediately tells you the type of the map if there is one, and throws the word `Map` at you if you weren't otherwise clear.
- Key and Value are as expected in a lot of languages, the first two arguments. Extend to tuples for the multi-key version, which is also familiar.
- Attributes like `max_entries` come last and are named (not positional). These are likely the least relevant part of the declaration, so last is good. Would be nice if these allowed trailing commas, but it isn't a dealbreaker.
- Leaves the door open for `= {};` initialisation, which could be a nice bit of sugar over filling them in in BEGIN manually.

The only thing I do wonder is whether we should reconsider using `->` to denote a mapping from key(s) to value. IMO in most languages which use generics for maps the comma makes a lot of sense, but in BpfScript maps are a first class feature. Is a bit of special syntax to demonstrate the "mapping" (or lack thereof if it's not there) a nice to have? I kind of like it.

My 2c.

**Comment by jordalgo** [comment]
> ```
> let @z : PercpuHashMap<int8, uint64,
>     max_entries = 64,
>     other_attr = "whatever",
> >;
> ```

@JakeHillion I don't think there is a technical reason against this syntax but I do have other reasons why I'm not a fan 😄  :
- It wouldn't align with probe attributes if we end up having those. I think coming up with a consistent "attribute" syntax for anything that needs them would help make scripts both easier to read and write. And I don't see us being able to re-use the above syntax for probes.
- In general, I find it visually unappealing and hard to read. In my mind (and this speaks to @viktormalik's question above about where we draw the line between an attribute and type information) the above declaration has 2 categories of information (map attributes which include the map "type" and the key/value types) that are mixed into three kinds of syntax. And to be fair, the same is true for what I suggested earlier e.g.
  ```
  [max_entries = 32]
  [other_attr = "whatever"]
  let @z: PercpuHashMap<int8, uint64>;
  ```
  Which is why I think we should NOT opt for adding the map type in front of the angle bracket (even though it's tempting). Let's keep attributes and key/value types distinct visually. This also gives the user the most flexibility in terms of declaring maps and utilizing any of these optional parts of the declaration (e.g. it feels weird to let them do `let @map: PercpuHashMap<>`). So, again, all of these would be legal:
  ```
   [type = LRUHashMap]
   [max_entries = 32]
   let @x <int, uint64>;

   [type = LRUHashMap]
   let @z: <int8, uint64>;
   
   [type = LRUHashMap]
   let @y;

   let @v: <int8, uint64>;

   let @u;
  ```

Note: I don't think I have a strong opinion yet on this:
```
[type = LRUHashMap]
[max_entries = 32]
let @x <(string, int8), uint64>;
```
vs.
```
[type = LRUHashMap;
 max_entries = 32;]
let @x <(string, int8), uint64>;
```
(maybe the former because it's easier to align the attributes visually).

**Comment by viktormalik** [comment]
> the above declaration has 2 categories of information (map attributes which include the map "type" and the key/value types) that are mixed into three kinds of syntax.
> [...]
> Which is why I think we should NOT opt for adding the map type in front of the angle bracket (even though it's tempting). Let's keep attributes and key/value types distinct visually.

@jordalgo this is a good argument and I agree that we should try not to mix different syntax kinds to express semantically similar things.

With that, I'm still in doubt whether key/value types are so distinct from other map attributes to get a different syntax (`<>` vs named attributes in `[]`).

My arguments:
- Map key/value types are not necessarily decoupled from the map type. For instance for array maps, you can't choose the key type, it's always `u32` so we'd only want the user to specify the value type. Other map types (e.g. bloom filter) do not even have keys, only values. So, changing the map type (via attributes) inherently influences the key/value definition.
- Other map attributes are not pairwise independent either. As @ajor mentioned, for an array map, we'll probably want the user to always specify the map size (`max_entries`) so adding `[type = ArrayMap]` means that the user *must* give `[max_entries = ...]` as well.
- I'm not even sure if we want to force users to always give both the key and value types at the declaration either. You may want to just define the value type and let bpftrace auto-deduce the key type for you.

So, I think the important question here is: what makes key/value types so special that they should have a different syntax from the rest of the map attributes? At the moment, I don't really see a reason to do that. I admit that it *feels* like key/value specify the map "type" so they should use the same syntax as scratch variable types but as seen above, the rest of the attributes are a part of the map "type" as well. So, I'd probably suggest to settle on a single syntax for all map attributes.

With the attributes syntax (which seems a bit preferred here), the above @jordalgo's examples would become:
```
 [type = LRUHashMap]
 [key = int; value = uint64]
 [max_entries = 32]
 let @x;

 [type = LRUHashMap]
 [key = int8; value = uint64]
 let @z;
 
 [type = LRUHashMap]
 let @y;

 [key = int8; value = uint64]
 let @v;

 let @u;
```

Also, these would give semantic analyser errors:
```
[type = ArrayMap]
[key = int8; value = int64]
[max_entries = 8]
let @x;  // cannot override key type for ArrayMap

[type = ArrayMap]
[value = int64]
let @x;  // ArrayMap requires the map_entries attribute
```

**Comment by jordalgo** [comment]
@viktormalik In light of those restrictions on the map key and value types, I agree with your suggestion of making everything an "attribute" with a single syntax. It's simple, it's readable, it easily allows every attribute to optional, AND this syntax will also work for probe attributes.

Small follow-up question: Should we restrict attributes to a single one per line (no semi-colon to divide e.g. `[key = int; value = uint64]`). I find it's easier to read with just one per line (might also be easier to parse) but maybe that's too restrictive. e.g.
```
[type = LRUHashMap]
[key = int]
[value = uint64]
[max_entries = 32]
let @x;
```

**Comment by viktormalik** [comment]
I find it too restrictive and too space-consuming. For instance, the typical use-case would be to define an array map and this:
```
[type = ArrayMap; value = uint64; max_entries = 8]
let @x;
```
IMHO reads just fine and doesn't consume much vertical space.

**Comment by fbs** [comment]
I don't really like mixing the type info into the attributes

- it doesn't seem consistent with how `let` does types for normal vars
- its a lot of additional typing to do
- the info doesn't seem relevant to the type system
- I don't see how you can make it work with automatic type detection. E.g. what if you only want to set `max_entries` on your array, do you then also have to specify
- 

What if instead of types and attributes, maps have an initializer

```
let @: Bloomfilter<i32> = Bloomfilter(k=10)
let @: LRUHashMap<i8, i8> = LRUHashMap(max=10)
```

There might be some nice tricks to automatically determine the type, making the type signature itself optional

**Comment by jordalgo** [comment]
> the info doesn't seem relevant to the type system

Yeah, I thought so too. This is why I had originally grouped these into two separate buckets: types for type system (key/value) and other "attributes" but given what @viktormalik and @ajor brought up, it feels like this separation isn't clean e.g. I think this syntax is very confusing: `let @: Bloomfilter<i32>` - is it the key type? is it the value type?

> What if instead of types and attributes, maps have an initializer
> `let @: LRUHashMap<i8, i8> = LRUHashMap(max=10)`

This doesn't look bad but I don't like that it closes the door on actual map initialization (e.g. if we wanted to insert keys and values at initialization, how would that look?) And to restate it's not consistent if/when we add probe attributes (https://github.com/bpftrace/bpftrace/issues/3360) - I think it would be really nice if we had one syntax for map and probe attributes.

**Comment by fbs** [comment]
> > the info doesn't seem relevant to the type system
> 
> Yeah, I thought so too. This is why I had originally grouped these into two separate buckets: types for type system (key/value) and other "attributes" but given what @viktormalik and @ajor brought up, it feels like this separation isn't clean e.g. I think this syntax is very confusing: `let @: Bloomfilter<i32>` - is it the key type? is it the value type?

A bloom filter doesn't have keys, right? The fact that everything is a map is just a bit confusing

> 
> > What if instead of types and attributes, maps have an initializer
> > `let @: LRUHashMap<i8, i8> = LRUHashMap(max=10)`
> 
> This doesn't look bad but I don't like that it closes the door on actual map initialization (e.g. if we wanted to insert keys and values at initialization, how would that look?) And to restate it's not consistent if/when we add probe attributes (#3360) - I think it would be really nice if we had one syntax for map and probe attributes.

Hmm I thought map initialization would still happen in begin probes, if needed. Although I agree that having `let @ = somefn()` in the 'global' bit also looks a bit weird and does imply that `let @ = [1,2]` should also work.

> 
> > I don't see how you can make it work with automatic type detection. E.g. what if you only want to set max_entries on your array, do you then also have to specify
> 
> Can you elaborate on this?

Say you have a program like:

```
[max_entries=8]
let @a 

kprobe:bla {
   @a[tid] = 1
}
```

In this case you need to run some kind of type inference before you validate the attribute, as the attribute won't work on any type.

but in 

```
[type=hash,max_entries=8]
let @a 

kprobe:bla {
   @a[tid] = 1
}
```

You need the attribute to be able to do type checking.

My worries is that it creates a lot of interdependencies between passes (or it becomes the type and attribute pass). Or you need ordering where you have passes that 1. extract type info from attributes, 2. run the type inference, 3. full attribute validation

**Comment by viktormalik** [comment]
> This doesn't look bad but I don't like that it closes the door on actual map initialization (e.g. if we wanted to insert keys and values at initialization, how would that look?) And to restate it's not consistent if/when we add probe attributes (#3360) - I think it would be really nice if we had one syntax for map and probe attributes.

I think that we should clarify one thing here. Map `type` (hash, array, ...) and `max_entries` are **not** attributes, they are a fundamental part of the map's type. It's the same as for "normal" compound types in programming languages - for instance in C, the array size is the part of the variable's type, not just some optional attribute. The fact that some of this information is optional in bpftrace (b/c it has a default value) doesn't make it decoupled from the underlying map type.

After giving it some thought, I'm starting to lean back towards using the map type in front of the angle bracket. Here's my reasoning:
- the map type is the fundamental information which **must** be defined for every BPF map,
- the map type specifies what other fields (attributes if you will) must be set on the map type - key/value type, max entries, additional, flags, etc.
- therefore, the `MapType<...>` syntax actually makes sense - `MapType` defines the actual data type (i.e. the layout of the memory storage for this map), the rest in `<...>` just defines additional attributes of that type.

In fact, it would be quite analogous (both syntax- and semantics-wise) to C++ template types - in `std::vector<int>`, `std::vector` says how this is stored in memory (contiguous sequence of elements) and `<int>` says what's the size of each element.

So I'd say that we should try to squeeze all the information into the angle brackets, although I agree that it won't be too nice when there will be many fields set. OTOH, I don't expect many users to specify all kinds of fields, most of them will use default values for most.

Some of the above examples would be:
```
// These two are equivalent
let @: Bloomfilter<int32, 10>;
let @: Bloomfilter<value=int32, max_entries=10>;

// Same here
let @: LRUHashMap<int8, int8, 10>;
let @: LRUHashMap<key=int8, value=int8, max_entries=10>;
```

For the default map type (`HashMap`), you could omit the first part and just give key/value types pair (which is IMO a typical use-case) so the declaration would be in line with scratch variable declarations.
```
let @: <int8, int8>;
```

> A bloom filter doesn't have keys, right? The fact that everything is a map is just a bit confusing

Yes, this is a great point. We call everything a "map" b/c kernel does the same but many of these just aren't maps. Bloom filter is one example but one could also argue the array map is not a map, it's an array (with standard indexation by an integer).

We may want to consider changing our docs to say that values starting with `@` are "global values/variables". It would also prevent some confusion like why scalar globals (`@x = 1`) are referred to as maps or why bloom filter (once we support it) doesn't have keys.

**Comment by fbs** [comment]
> I think that we should clarify one thing here. Map `type` (hash, array, ...) and `max_entries` are **not** attributes, they are a fundamental part of the map's type. It's the same as for "normal" compound types in programming languages - for instance in C, the array size is the part of the variable's type, not just some optional attribute. The fact that some of this information is optional in bpftrace (b/c it has a default value) doesn't make it decoupled from the underlying map type.
> 

Hmm, I guess arrays are a bit of a special case as pre-allocated static size arrays are possible, which can be useful into for static checking. For a hash or bloom filter this info seems a lot less relevant to the type system. E.g. for a bloom filter you'd specify the number of items and the acceptable false positive rate, but thats not  relevant to the type of data you put in it, which is the bit the type checking cares about.

> We may want to consider changing our docs to say that values starting with @ are "global values/variables". It would also prevent some confusion like why scalar globals (@x = 1) are referred to as maps or why bloom filter (once we support it) doesn't have keys.

+1

**Comment by jordalgo** [comment]
> My worries is that it creates a lot of interdependencies between passes (or it becomes the type and attribute pass). Or you need ordering where you have passes that 1. extract type info from attributes, 2. run the type inference, 3. full attribute validation

I feel like we're kinda in that boat already with all the back type propagation stuff we need e.g. https://github.com/bpftrace/bpftrace/pull/3536 - I'm not saying that's a good reason but inferring the map "type" based on how it's used seems inline with the flexibility we already provide to users.

> We may want to consider changing our docs to say that values starting with @ are "global values/variables". 

I'm for this idea. It's always been the source of a lot of user confusion. And maybe we can update the error messages to stop calling things "scalar maps".

As far as this syntax goes:

```
// These two are equivalent
let @: Bloomfilter<int32, 10>;
let @: Bloomfilter<value=int32, max_entries=10>;
```
For me the only way to NOT make this super confusing to users is to force named arguments and not let anything be positional. There is no way I wouldn't have to look at the docs every time I saw this: `let @: Bloomfilter<int32, 10>;`. The mixing of types and values without being explicit makes my eyes water like I'm peeling an onion 😿 
I think this restriction is ok though because other than users saying they want to set this **global** to be a Bloomfilter /LRUHashMap we haven't gotten any (IIRC) requests to be able to state the key/value types up front. So the majority of the time we'll probably only see `let @: Bloomfilter;` (we don't need the angle brackets) or `let @: Bloomfilter<max_entries=10>;`. Type inference for the keys (if they exist) and values should still work as normal.

Forcing named attributes also makes the multi-line definitions a little less hideous (though I still don't like it):
```
let @z : PercpuHashMap<
    key=int8,
    value=uint64,
    max_entries = 64,
    other_attr = "whatever",
>;
```

And to keep sounding like a broken record, the above syntax (whether or not we are considering them "attributes" or fundamental parts of the map's type) will be different from the syntax we provide for probe "attributes". I don't love forcing users to learn both even though their use is going to be minimal.

**Comment by viktormalik** [comment]
> For me the only way to NOT make this super confusing to users is to force named arguments and not let anything be positional. 

Sure, no problem with that on my side. As we already agreed, terseness is not a target for this syntax so requiring explicit names for attributes sounds good to me.

> Forcing named attributes also makes the multi-line definitions a little less hideous (though I still don't like it):
> 
> ```
> let @z : PercpuHashMap<
>     key=int8,
>     value=uint64,
>     max_entries = 64,
>     other_attr = "whatever",
> >;
> ```

I'm not saying it looks nice but it's not much worse than what we had with many attributes above. And, as you already mentioned, we don't expect users to set this many attributes often.

> And to keep sounding like a broken record, the above syntax (whether or not we are considering them "attributes" or fundamental parts of the map's type) will be different from the syntax we provide for probe "attributes". I don't love forcing users to learn both even though their use is going to be minimal.

I don't think we have too many good options here. I feel like we all agree that at least the key/value types should be set the same way as scratch variable types which means that we're introducing a new syntax (`<>`) anyways. Therefore, I think that we should put everything related to the map type into the same place b/c in my opinion (and this also addresses @fbs comment above) the map type should not just be relevant to bpftrace type system but it should also be relevant to users to understand (and influence) how the map is stored in memory. And the number of entries is a part of that.

With that, I still see space for adding attributes to maps in future. For instance, we may have an attribute which influences how integers are printed for that particular map when using `print` (in decimal or hex format).

**Comment by jordalgo** [comment]
Ok, it feels like we're converging around this syntax (correct me if I'm wrong):
```
let @x: BloomFilter;

let @y: PercpuHashMap<max_entries = 64>;

let @z : PercpuHashMap<
    key=int8,
    value=uint64,
    max_entries = 64,
    other_attr = "whatever",
>;
```

And just to summarize a bit why:
- It's consistent with scratch variable declarations (as per @JakeHillion and @fbs point above)
- It still leaves the door open for global "attributes" (e.g. how to print or sort perhaps) that would be consistent with potential probe "attributes"
- Everything to the right of the colon is part of the global variable's definition or "type" (as per @viktormalik 's point)
- It leaves the door open for `= {};` initialization
- The syntax makes it immediately clear what kind of global variable it is: BloomFilter, HashMap, etc. and we can use that when validating the use of it in the script

Please continue commenting with concerns/suggestions. Just thought it was good to get a quick summary of where we, hopefully, are.

**Comment by fbs** [comment]
yes seems reasonable. A fairly common way of doing this.

Will the names be optional? I.e. is `HashMap<int8, int8>` equal to `HashMap<key=int8, value=int8>`?

I'm not yet convinced about having `max_entries` there. To me thats a probe attribute of sorts and not related to the type but do like that you dont have to slap on on a mixture of attributes and type hints.

**Comment by jordalgo** [comment]
> Will the names be optional?

No. All names are required so this would not be valid: `HashMap<int8, int8>` or this `let @: Bloomfilter<int32, 10>;`.

> I'm not yet convinced about having max_entries there.

I think @viktormalik made a good point that in C the number of elements is part of the type syntax e.g. `int myArr[20]`.

**Comment by fbs** [comment]
> > I'm not yet convinced about having max_entries there.
> 
> I think @viktormalik made a good point that in C the number of elements is part of the type syntax e.g. `int myArr[20]`.

To me thats a special case that as knowing the array size at compile time can help as you can allocate a fixed block of stack space (could also argue that `int[] arr[8]` would've make more sense for C). Knowing the amount of entries in a hashmap seems a lot less relevant and brings a lot of runtime info into the type system.
It also makes the type signatures very complex, a combination of attributes and types seems like a better fit to me (or more object like initialization)

If we do this:

```
[max_entries=1000]
let @map: Bloomfilter<int64>
```

The types and other attributes are clearly separated. It also keeps the type signature simple to type, don't have to add `value=int64` , but that can be supported

**Comment by danobi** [comment]
## High level
This feature is big enough of a carrot and wide enough of a design space to consider marking this as an unstable feature (https://github.com/bpftrace/bpftrace/issues/3521). We've had enough demand for this that users will go out of their way to deal with unstable-ness. We've seen this in practice with big strings recently.

Marking as unstable lets us collect feedback / gain experience _and_ continue discussion. The bar for stabilizing this feature should be at least as high as an upfront design discussion.

## Design thoughts

Based on brief discussion in https://github.com/bpftrace/bpftrace/issues/3360, my feeling was that attributes are a way to specialize behavior. So I was thinking about attributes more in the context of specializing maps to percpu, arraymap, lru, etc. Bloom filter is a good example but a borderline case. In theory existing map syntax can work with bloom filter (although in practice the bpf-side helpers are different, see https://docs.kernel.org/bpf/map_bloom_filter.html). Obviously we are not going to support arena, structops, prog_array, etc. with this syntax, right? For the latter, I expect we'd come up with a different API.

When I think of map attributes as a specialization, attributes make a lot of sense to me. Composable, optional, and explicit. It also fits in nicely with probe-level attributes. Less syntax is better - every additional construct is a new "weird" thing for user to learn. Better to reuse existing constructs where we can.

IOW, I don't see this as much of a typing decision. Bringing the scope of this design to the level of typing complicates things. And I don't see it as necessary.

**Comment by jordalgo** [comment]
> In theory existing map syntax can work with bloom filter ... Obviously we are not going to support arena, structops, prog_array, etc. with this syntax, right?

@danobi  I think one of the things that came out of this discussion (and please feel free to disagree) is that the `@` sigil is used to describe "globals" and not necessarily just maps, which has maybe caused the scope of this discussion to grow e.g. the discussion of bloom filter.

It does seem like we're having a hard time agreeing on where the line is between what is an "attribute" and part of the global's "type".

If we carry the idea further that `@` describes a global then I think, to keep consistent with scratch variable type declarations (as well as future function arguments), we need to support some kind of "type" to the right of these globals e.g.

```
let @a: LRUHashMap;
let @b: BloomFilter;
let @c: Arena;
```

This feels to me like a good baseline. Now the question becomes, where do we put the other information about these globals. 

It also seems, from the discussion above, that we're not considering NOT having (sorry for the double negative) the "attribute" syntax e.g. `[myattr = whatever]` as this will be used by probes as well. So that leaves us with this:

```
[myattr = whatever]
let @a: LRUHashMap;
```

So the question then becomes: where do we put the inner "type" information about this global? TBH, I'm not sure if I care at this point because I don't think we have a good use-case for users actually utilizing this feature. It would mostly be for further communication to the reader of a script as we can infer the inner type information as normal - and so far we haven't had much of a problem with type ambiguity. 

My vote at the moment, though it seems to change daily, is to just allow users to specify the key and value types for map globals in the attribute syntax. This more-or-less allows us to put off the decision to add an additional angel bracket syntax. Even if we decide to add one later, we could always still support setting the "type" in the attribute syntax. I don't think it costs us anything. So...

```
[myattr = whatever]
[key = int64]
[value = string]
let @a: LRUHashMap;
```

**Comment by jordalgo** [comment]
> Marking as unstable lets us collect feedback / gain experience and continue discussion. The bar for stabilizing this feature should be at least as high as an upfront design discussion.

Also totally good for marking this feature as unstable/experimental.

**Comment by danobi** [comment]
> > In theory existing map syntax can work with bloom filter ... Obviously we are not going to support arena, structops, prog_array, etc. with this syntax, right?
> 
> @danobi I think one of the things that came out of this discussion (and please feel free to disagree) is that the `@` sigil is used to describe "globals" and not necessarily just maps, which has maybe caused the scope of this discussion to grow e.g. the discussion of bloom filter.

Hmm, I'd be careful there. `@` is more than just scoping - it also enables kev/value semantics. ie `@[key] = val`. Then one could wonder why they can't do `$scratch[key] = val`.

Not saying it shouldn't be considered. But I recommend spinning that off into its own discussion.

> 
> It does seem like we're having a hard time agreeing on where the line is between what is an "attribute" and part of the global's "type".
> 
> If we carry the idea further that `@` describes a global then I think, to keep consistent with scratch variable type declarations (as well as future function arguments), we need to support some kind of "type" to the right of these globals e.g.
> 
> ```
> let @a: LRUHashMap;
> let @b: BloomFilter;
> let @c: Arena;
> ```
> 
> This feels to me like a good baseline. Now the question becomes, where do we put the other information about these globals.

Thanks for laying this out. I had a good sleep on this and have some new thoughts. Looks like I didn't think enough and what I was proposing would just complicate things in the future.

Originally I was hoping we could get away with the minimal amount of changes. For completeness, here's what I had in mind:

```
[key_type = int64]
[value_type = int64]
[map_type = percpu]
@[$foo] = $bar;
```

So skip `let` stuff altogether but not rule it out for future.

But then I had the thought that it'll be important that users can pass a map to a user defined function. In that case, we need the map_type/key_type/value_type information in the type system in order for UDF to be generic across maps. This will be crucial for a bpftrace standard library as well as the C plugin mechanism. For example, we'll want to be able to do this:

```
fn top_n(@m: PercpuHash<i64, i64>, $n: int64) {
  // do stuff with @m 
  ...
}
```

It's important to `top_n` that it knows `@m` is percpu as well as what the key/value types are. So that different maps can be passed into the same function. But it's **not** important it knows max_entries, numa affinity, etc. So basically what @fbs is proposing earlier.

> 
> It also seems, from the discussion above, that we're not considering NOT having (sorry for the double negative) the "attribute" syntax e.g. `[myattr = whatever]` as this will be used by probes as well. So that leaves us with this:
> 
> ```
> [myattr = whatever]
> let @a: LRUHashMap;
> ```
> 
> So the question then becomes: where do we put the inner "type" information about this global? TBH, I'm not sure if I care at this point because I don't think we have a good use-case for users actually utilizing this feature. It would mostly be for further communication to the reader of a script as we can infer the inner type information as normal - and so far we haven't had much of a problem with type ambiguity.

Previously agreed. But see UDF example above.

**Comment by jordalgo** [comment]
> . In that case, we need the map_type/key_type/value_type information in the type system in order for UDF to be generic across maps.

Very good point and sad I missed it myself lol. 

> So skip let stuff altogether but not rule it out for future.

Don't agree with this. We added `let` after a lot of discussion and though it seems a bit superfluous in this case, it's an indication of declaration be they scratch variables, globals, maps, whatever. I think it's inconsistent if we don't use `let` no matter what we're calling things defined with `@`.

**Comment by danobi** [comment]
One other consideration: if we implement map-in-map (https://github.com/bpftrace/bpftrace/issues/3553), we'll need a way to express inner map type as well. So having the key/value argument be present in the type name would ensure map-in-map can be captured in type system. For example:

```
[myattr = whatever]
let @m: Hash<i64, Hash<i64, i64>>;
```

**Comment by viktormalik** [comment]
> > . In that case, we need the map_type/key_type/value_type information in the type system in order for UDF to be generic across maps.
> 
> Very good point and sad I missed it myself lol.

Yeah, completely forgot about user-defined functions.

> It's important to top_n that it knows @m is percpu as well as what the key/value types are. So that different maps can be passed into the same function.

Yes, this looks like a reasonable distinction point - whatever information is needed to be passed to functions must be a part of the map type, the rest can be attributes.

> But it's not important it knows max_entries, numa affinity, etc.

Isn't it? I'm especially looking into `ArrayMap` here. I'm not sure, but I'd expect that the verifier checks bounds when accessing the map so it would fail on OOB access. Since we don't want to expose users to verifier errors, we'll probably need to make `max_entries` mandatory and therefore somehow pass it to UDF.

> One other consideration: if we implement map-in-map (#3553), we'll need a way to express inner map type as well. So having the key/value argument be present in the type name would ensure map-in-map can be captured in type system. For example:
> 
> ```
> [myattr = whatever]
> let @m: Hash<i64, Hash<i64, i64>>;
> ```

Hmm, can the outer and inner maps have different attributes (like `max_entries`)? If yes, then this syntax won't allow to express them.

**Comment by danobi** [comment]
> Isn't it? I'm especially looking into ArrayMap here. I'm not sure, but I'd expect that the verifier checks bounds when accessing the map so it would fail on OOB access. Since we don't want to expose users to verifier errors, we'll probably need to make max_entries mandatory and therefore somehow pass it to UDF.

But array maps still use bpf_map_elem_lookup() API, so the bounds check is done there. Accessing out of bounds results in a zero-d out value. 

> Hmm, can the outer and inner maps have different attributes (like max_entries)? If yes, then this syntax won't allow to express them.

Yeah, good point. Unless we added `inner_` prefix to attributes. Like `inner_map_map_entries`. Kinda ugly. Doesn't fully solve arbitrary map-in-map nesting if kernel ever supports that. No other ideas from me here.

**Comment by jordalgo** [comment]
> Yeah, good point. Unless we added inner_ prefix to attributes. Like inner_map_map_entries. Kinda ugly. Doesn't fully solve arbitrary map-in-map nesting if kernel ever supports that. No other ideas from me here.

I also was thinking of some named attribute like this (I agree it's ugly). I also don't see UDFs needing to specify the `max_entries` for a map type signature e.g. `fn(@mp: Hash<i64, Hash<i64, i64>>;): int64`. Another possibility is to force map definitions to only be single level if you want/need to add attributes.

```
[myattr = whatever]
let type MyInnerMap = Hash<int64, int64>;

[myattr = whatever]
let @b: Hash<i64, MyInnerMap>;
```

**Comment by viktormalik** [comment]
> But array maps still use bpf_map_elem_lookup() API, so the bounds check is done there. Accessing out of bounds results in a zero-d out value.

Right, forgot about that. Then it's ok and we don't need to pass the size to functions.

> I also was thinking of some named attribute like this (I agree it's ugly). I also don't see UDFs needing to specify the `max_entries` for a map type signature e.g. `fn(@mp: Hash<i64, Hash<i64, i64>>;): int64`. Another possibility is to force map definitions to only be single level if you want/need to add attributes.
> 
> ```
> [myattr = whatever]
> let type MyInnerMap = Hash<int64, int64>;
> 
> [myattr = whatever]
> let @b: Hash<i64, MyInnerMap>;
> ```

This is a good idea. I don't think adding attributes would be used very often (especially for nested maps) and this provides a consistent way to specify them for arbitrary map nesting.

It looks like you've convinced me - as long as the map types are specified as `Hash<int64, int64>`, I'm ok with specifying the rest of the fields via attributes.

**Comment by jordalgo** [comment]
Ok folks, just to summarize from office hours and the above discussion. We're converging on this syntax. Additionally we will be putting this functionality behind a feature gate (https://github.com/bpftrace/bpftrace/issues/3521)

All of these below will be valid:

```
let @a;

let @b: Hash<u64, u64>;

[myattr = whatever]
let @c: Hash<u64, u64>;
```

And when we add user-defined types and maps in maps:
```
[myattr = whatever]
type MyInnerMap = Hash<int64, int64>;

[myattr = whatever]
let @b: Hash<i64, MyInnerMap>;
```

Additionally, we need to figure out what BPF map types are valid and what types can be used with what internal and external types.

Please thumbs up if you agree with this decision and thumbs down if you don't (but please add reasons why not in further comments below).

**Comment by jordalgo** [comment]
Quick maintainers roll call. If you disagree with any part of the above proposal, please comment soon as I plan to start development work in the next few weeks. @fbs @ajor @danobi @viktormalik

**Comment by amscanne** [comment]
Declaration syntax aside, was there closure on where the statements are placed? Just as a datapoint, the idea that these lets exist at the global scope (alongside probes) was non-obvious to me, and I guess I'm just wondering why they couldn't be used in any probe (mirroring how maps are already "declared" implicitly). This would simplify one-liners (since the let would appear inline in the single probe), and you could just have a convention to declare all maps up front in the `BEGIN` probe for complex programs.

Then again, an argument in favor of having the statement at the top-level would be if there as also an intent to allow global variables of non-map types (which could be implemented using a suitable lift into a map for the set of scalar globals), which would fit naturally with this. I would probably try a global `let $x` as soon as I discovered the global-level map variables. It's worth noting that this does open *another* can of worms, as I also want to be able to declare probe-local maps (shared state just between invocations of the probe, but not within the global namespace) ;) Is there a sense for this roadmap?

**Comment by jordalgo** [comment]
> was there closure on where the statements are placed

@amscanne Thanks for following up on this. In my mind, when using the `let` syntax you are declaring a scope that this variable is valid in. Since `@` variables are not reset/recreated during every probe entry, utilizing the let syntax to define the their "type" and additional attributes should happen at the global level. This doesn't prevent users from just using globals as they have been - that being said maybe we should think about allowing globals to specify their "type" inside a probe for the one liner case (as you mentioned) e.g.
```
BEGIN { @a: LRUHash = 1; }
```

>  if there as also an intent to allow global variables of non-map types

We've talked about this case and that's why I'm careful to say "global" above and not map 😉 

> as I also want to be able to declare probe-local maps (shared state just between invocations of the probe, but not within the global namespace) 

What is the use case for this? Why couldn't the script just not use/reference said global in other probes?

**Comment by amscanne** [comment]
> What is the use case for this? Why couldn't the script just not use/reference said global in other probes?

I think it's pretty easy to contrive scenarios where implicit globals lead to paper cuts. Any case where you *think* you're making a local change, but you're actually affecting something globally. Suppose that I had two probes, `X` and `Y`, and maybe they tracked some invocation time using `@since` and `@last`. Maybe I go in and "clean up" `X`, and change the map used to `@last`. Now I'm scratching my head wondering why `X` and `Y` are producing occasionally garbage results (they are both using the same map now, but I only changed `X`!). Having an explicit `let @last` in each probe would make it clear that it was only local to that specific probe, and changes to naming there don't affect any other probes implicitly.

I don't think this is urgent or anything, I'm just curious about the long-term trajectory. Having globals/locals be dependent on the scope, and maps/scalar values a function of the type (with both supported in both scopes) makes a ton of sense, and just that today you have only global maps, local scalars. Any implicit declarations would continue to default to those scopes for backwards compatibility, but as programs grow you encourage explicit declarations (presumably).

**Comment by jordalgo** [comment]
The issue I see with allowing persistent maps (or anything) local to only specific probes is perhaps in the declaration - it feels confusing that a data structure is persisting even though every time the program is called (like a function) you are re-declaring it with `let`, which says to me this variable is new/empty at this point forward. I'm not opposed to the concept of probe-specific persistent data structures, I just don't think the `let` syntax is suited for them. Maybe having file-local maps/globals when we start allowing imports of other bpftrace scripts is a good first start OR we could have an attribute that allows you to specify in which probes it can be accessed (though that feels too heavy weight).

**Comment by jordalgo** [comment]
In some discussions this past week, there was a general consensus (among @amscanne, @danobi, @4ast, @tyroguru, @ttreyer, and myself) that allowing and/or forcing users to specify types for scratch variables, globals/maps, and user defined functions has multiple issues, at least currently, which @danobi also mentioned in the linked comment above:
- there is a non trivial amount of work to remove hard-coded bpftrace types in the lexer/parser or, to put it more simply, to support both the range of all C types AND internal bpftrace types in the syntax
- bpftrace is already a statically typed but fully type inferred language so why force users to deal with types more directly
- it goes against our bpftrace [language guidelines](https://docs.google.com/document/d/1a5vla623n0Z9oaUCNLaaDux6X6FO4qYfBs9zLJ6nzhA/edit?tab=t.0#heading=h.x3ejk850hlgl) to keep the language terse

This is obviously still open for discussion but given the above, I want to propose a simplified version of what was already agreed upon [above](https://github.com/bpftrace/bpftrace/issues/2954#issuecomment-2463608135):
```
let @a; // valid but useless until we have something like scoped maps

[myattr = val]
let @b;

// more specifically
[type = LRUHash,
 max_entries = 250]
let @b;

// OR
[type = LRUHash]
[max_entries = 250]
let @b;
```

This basically just removes the type signature and moves the map "kind" or "type" to become an "attribute". Additionally, we would start by limiting the map attributes to just "type" and "max_entries" (we might not want to go beyond that anyway). Since we had already agreed upon map "attributes", this feels reasonable but feedback is always welcome.

**Comment by 4ast** [comment]
> [type = LRUHash, max_entries = 250]
> [type = LRUHash] [max_entries = 250]

I would pick one syntax instead of implementing and testing two. I have slight preference to former.

**Comment by fbs** [comment]
> There are issues with this design. The verifier considers more than key/value types as relevant type information for maps.

But is bpftrace also going to use that type information to do anything? It's not clear to me why the verifier doing that should impact syntax.

With this syntax you seem to get stuck if its anything other than a simple map declaration, which might hurt future features (e.g. map of maps).

An initializer function might be nicer, it won't solve the above, but it doesn't block anything either. It fits the 'things are typed on first use' that rest also has

```
let @b = LRUHash(max_entries=20)
```

**Comment by danobi** [comment]
>> There are issues with this design. The verifier considers more than key/value types as relevant type information for maps.
>
> But is bpftrace also going to use that type information to do anything? It's not clear to me why the verifier doing that should impact syntax.

The main reason (IMO) that the design went this direction (neither I nor Viktor were particularly convinced of the template syntax IIRC) was to support UDF. For UDFs, we needed to match verifier semantics on subprog argument matching. This was b/c we wanted to accept maps as function args. We now know our type system would have been incomplete.

From going through that exercise, I'm increasingly of the opinion that we should only extend bpftrace language when it provides overwhelming benefit. Certainly not for one-off use cases. Bpfscript extensions are dangerously complex due to the ad-hoc nature of the original design. Adin has some really good points above about unexpected complexity.

For example for your initializer proposal, we'd at minimum need to add named parameters. Which requires figuring out how we deal with new helpers going forwards. And if we need to add named parameters for older helpers. And if name in parameter is optional. Additionally, it's kind of backwards from the map helpers. eg `@ = count()` does something new each call. But `let @ = LRUHash(..)` is an only-once?

Not trying to pick on it too much - just trying to make a point that we could accidentally add way more complexity to the system than we bargained for. And even if we chased all these loose ends down, we still only unlocked a single feature.

**Comment by ajor** [comment]
>>> There are issues with this design. The verifier considers more than key/value types as relevant type information for maps.
>>
>> But is bpftrace also going to use that type information to do anything? It's not clear to me why the verifier doing that should impact syntax.
>
> For UDFs, we needed to match verifier semantics on subprog argument matching. This was b/c we wanted to accept maps as function args. We now know our type system would have been incomplete.

Types are meant to be useful from a user's point of view - attributes like max keys don't necessarily have to make map types distinct at the BpfScript level, even if they do at the BPF level. If a user's written a function which expects to take a hash map of ints to ints, they most likely don't care whether the passed-in map has space for 4096 elements or 10000 elements.

The compiler can do work behind the scenes to fix up any disconnect between the BpfScript type system and what the verifier wants. The simplest way to resolve it is to just inline all functions instead of using BPF subprogs. With subprograms, generating specialised programs for each function call with verifier-distinct map types might be a solution (similar to what we do with probe expansion).

**Comment by fbs** [comment]
> > With this syntax you seem to get stuck if its anything other than a simple map declaration, which might hurt future features (e.g. map of maps).
> 
> I'm not sure if I understand how it hurts future features. Even for maps in maps you could do something like this:
> 

More that you can't have the 'recursion' that usual type notation has to specify inner types. E.g a `array map where the inner type is a hashmap string[16]->int64` 

> I don't want to get stuck in analysis paralysis here but maybe dragging our feet on this particular syntax isn't terrible 

Fair, removing attributes later on isn't terrible.



> > > There are issues with this design. The verifier considers more than key/value types as relevant type information for maps.
> > 
> > 
> > But is bpftrace also going to use that type information to do anything? It's not clear to me why the verifier doing that should impact syntax.
> 

> For example for your initializer proposal, we'd at minimum need to add named parameters. 

I dont think its a requirement. It will depend on how many args an initalizer would have to have.

However specifying the key type of an array with this is not going to work, unless you get types in there which is worse. So this is not a full alternative.


> [..] But `let @ = LRUHash(..)` is an only-once?

Fair but as long as its only allowed in a global section that doesn't seem to confusing.

It could also be re-usable as dynamic map creation, but that would need a good usecse first 

> 
> Not trying to pick on it too much - just trying to make a point that we could accidentally add way more complexity to the system than we bargained for. And even if we chased all these loose ends down, we still only unlocked a single feature.

True, there doesn't seem to be a great solution for this.

**Comment by viktormalik** [comment]
> In some discussions this past week, there was a general consensus (among [@amscanne](https://github.com/amscanne), [@danobi](https://github.com/danobi), [@4ast](https://github.com/4ast), [@tyroguru](https://github.com/tyroguru), [@ttreyer](https://github.com/ttreyer), and myself) that allowing and/or forcing users to specify types for scratch variables, globals/maps, and user defined functions has multiple issues, at least currently, which [@danobi](https://github.com/danobi) also mentioned in the linked comment above:
> 
> * there is a non trivial amount of work to remove hard-coded bpftrace types in the lexer/parser or, to put it more simply, to support both the range of all C types AND internal bpftrace types in the syntax
> * bpftrace is already a statically typed but fully type inferred language so why force users to deal with types more directly
> * it goes against our bpftrace [language guidelines](https://docs.google.com/document/d/1a5vla623n0Z9oaUCNLaaDux6X6FO4qYfBs9zLJ6nzhA/edit?tab=t.0#heading=h.x3ejk850hlgl) to keep the language terse
> 
> This is obviously still open for discussion but given the above, I want to propose a simplified version of what was already agreed upon [above](https://github.com/bpftrace/bpftrace/issues/2954#issuecomment-2463608135):
> [...]
> This basically just removes the type signature and moves the map "kind" or "type" to become an "attribute". Additionally, we would start by limiting the map attributes to just "type" and "max_entries" (we might not want to go beyond that anyway). Since we had already agreed upon map "attributes", this feels reasonable but feedback is always welcome.


Does this mean that we will not allow users to declare the key/value data type for maps? I was under the impression that this was the original motivation for this proposal. If not, the attributes make the most sense. I don't love them either but don't think we have anything better at hand.

> > But is bpftrace also going to use that type information to do anything? It's not clear to me why the verifier doing that should impact syntax.

> The main reason (IMO) that the design went this direction (neither I nor Viktor were particularly convinced of the template syntax IIRC) was to support UDF. For UDFs, we needed to match verifier semantics on subprog argument matching. This was b/c we wanted to accept maps as function args. We now know our type system would have been incomplete.

IMO, we originally went this direction b/c we anticipated bpftrace to do type checks for the specified key/value types which naturally made them a part of the "map type" (hence the "template" syntax). Then, UDFs jumped and we started thinking about aligning with what verifier does (which is probably not a good thing to do as correctly pointed out multiple times in this thread). As I wrote in https://github.com/bpftrace/bpftrace/issues/2516#issuecomment-2647793690, I agree that passing maps to UDFs should no longer be a valid use-case when thinking about this syntax as it has too many issues.

**Comment by 4ast** [comment]
> they most likely don't care whether the passed-in map has space for 4096 elements or 10000 elements

No. It's the other way around. map's max size (max_entries) is the main parameter of the map that users care about,
since it dictates the context where the map can be used. The tools (bpftrace and everything else) cannot predict this value.
Only the author of the script knows the upper bound of the map size. It might be 1 and it might be 1 million.

**Comment by jordalgo** [comment]
> Does this mean that we will not allow users to declare the key/value data type for maps?

That's correct but only for the time being. From my understanding, we really just want a way for users to supply map attributes (e.g. type, like LRU_HASH, and max_entries) so the key/value data types are not required. 

Granted, users already have to deal with C types for casting and whatnot, but IMHO we need to think more before moving away from inferred typing - I think it's one of bpftrace's selling points, part of what keeps the language terse, and users haven't been requesting a need to specify types (IIRC). I think because some of the code for User Defined Functions had already been merged I thought we were required to go down this path but now I don't believe that to be true. 

There are obviously good arguments to be made for making types more explicit in a language (e.g. typescript, Hack) but I don't think it's required to address the specific requests we've had from users. Again, not against making types explicit on principal for bpftrace, I just think we can delay it until things are more ironed out.

**Comment by viktormalik** [comment]
> Granted, users already have to deal with C types for casting and whatnot, but IMHO we need to think more before moving away from inferred typing - I think it's one of bpftrace's selling points, part of what keeps the language terse, and users haven't been requesting a need to specify types (IIRC). I think because some of the code for User Defined Functions had already been merged I thought we were required to go down this path but now I don't believe that to be true.
> 
> There are obviously good arguments to be made for making types more explicit in a language (e.g. typescript, Hack) but I don't think it's required to address the specific requests we've had from users. Again, not against making types explicit on principal for bpftrace, I just think we can delay it until things are more ironed out.

I'm not saying we should move away from inferred typing, I just thought that there was an ask to enable map declarations so that users can declare what kinds of maps they want to use and have bpftrace check that they are used (assigned to) correctly. If that is not the case at the moment, then it makes perfect sense not to add type declarations now and stick with attributes.

What we have to keep in mind, though, is that once such a request comes and we want to implement it, adding key/value types to attributes will mean that we will have multiple different ways of specifying types for the type system - one for scratch variables and function parameters (if those are ever implemented) and one for maps. That could be quite confusing.

**Comment by viktormalik** [comment]
> > they most likely don't care whether the passed-in map has space for 4096 elements or 10000 elements
> 
> No. It's the other way around. map's max size (max_entries) is the main parameter of the map that users care about, since it dictates the context where the map can be used. The tools (bpftrace and everything else) cannot predict this value. Only the author of the script knows the upper bound of the map size. It might be 1 and it might be 1 million.

I'd say that in bpftrace, it's both. A large portion of users will not care and be satisfied with whatever size is the default (they just want a key-value storage for a fairly small number of elements). The advanced users will want to specify the map size and will care about the context (functions/macros) where it can be used. That's why, IMO, `max_entries` should be an optional part of the map type.

**Comment by 4ast** [comment]
> > > they most likely don't care whether the passed-in map has space for 4096 elements or 10000 elements
> > 
> > 
> > No. It's the other way around. map's max size (max_entries) is the main parameter of the map that users care about, since it dictates the context where the map can be used. The tools (bpftrace and everything else) cannot predict this value. Only the author of the script knows the upper bound of the map size. It might be 1 and it might be 1 million.
> 
> I'd say that in bpftrace, it's both. A large portion of users will not care and be satisfied with whatever size is the default (they just want a key-value storage for a fairly small number of elements). The advanced users will want to specify the map size and will care about the context (functions/macros) where it can be used. That's why, IMO, `max_entries` should be an optional part of the map type.

That was a serious mistake in bpftrace design. It made users believe that bpftrace's map is a key/value storage like std::map or similar containers in all normal languages, but it is not. It's a fixed size hash table. It's not resizable. It won't allow more insertion after reaching the limit. Nothing like this exists in c++, python, awk, bash. It's a foot gun. Users that only need one or few elements in a hashmap are wasting all that preallocated unswappable kernel memory. While other users hit the limit and have their events dropped. The size should have been an explicit and mandatory parameter.

**Comment by fbs** [comment]
Based on the discussions above I still prefer an initializer function over attributes for `max_entries` specifically. Yes, without named function arguments its not great, however I do not expect many arguments for map creation.

So:

```
let @a = HashMap(/*max_entries=*/100)
```

Restriction is that this can only be used in the 'global' section, outside any probe ( at least for now?). It also removes the need to have a `max_entries` and a `type` attribute. Key/value types still have to be interred.

To me it reads as  `create a HashMap of size 100` and seems somewhat intuitive. More so than attributes for this use case.

**Comment by amscanne** [comment]
Given the level of discussion, one possible approach could be to defer some of the richer conversation and decisions around type specifications, annotations and even `let` scoping and do something very simple to unblock the known user needs enumerated at the beginning. This would allow the internal details to be fleshed out, and give breathing room for the language design aspects (perhaps as part of a cohesive effort) without having them in the critical path for Use Case 2 (map settings) and Use Case 3 (map types).

Specifically, you could do an even simpler version of what @fbs is proposing. This would not be a special/new initializer, but just using the existing inference model and existing syntax. A rough sketch would be:

* Introduce new builtins that are consumed exclusively for type inference. Consider `@foo = arraymap(10)`, `@bar = lruhashmap($3)`, `@baz = percpuarraymap(1)`. There are probably just the five map types that you care about for now (but obviously could start even smaller). The argument would specify the size and need to be an integer literal or parameter, like many other builtins.
* These statements would be consumed within the semantic analyzer to define a map type and size. Obviously, it would check for conflicts but the statements could otherwise appear anywhere, just like any inference. The key and value types would be inferred exactly as before (thought obviously there are additional checks depending on the map type).
* You could emit a `WARNING` when this is used that it is experimental and may disappear at any time.
* Although there are definitely additional challenges (part of the value of unblocking the fleshing out of details!), I believe that this approach even extends to map-of-maps without the need for specialized syntax. Consider `BEGIN { @foo = arraymap(2); @bar = hashmap(4); @baz = hashmap(4); @foo[0] = @foo; @foo[1] = @baz }`.

Such an approach lets you completely avoid any explicit type specification and avoids the need for immediate decisions around new syntax, required scoping of lets, etc. Maps are already global and declared implicitly, and the key and value types are already inferred, so I think this fits well with the current ergonomics. (And given no fundamental syntax changes and potential use of a `WARNING` for experimental features, it seems like a low risk approach.)

One variation would be to split out the inference of different aspects with separate builtins, e.g. provide a builtin `reserve` function or similar to allow users to control the capacity if they just want the default map. (Or perhaps a `setpercpu` or `setlru` for the same effect.) It makes some intuitive sense with one-liners like: `i:s:1 { reserve(@foo, 100); @foo[...] = ... }`, where users may not care about map types but *do* need to care about the capacity (as noted above).

**Comment by fbs** [comment]
+1

Having map 'initializers' (even if its internally not treated as a function, from bpfscript it still reads as a function) in `BEGIN` only could also work. So `BEGIN { @a = HashMap(10) }`. This leaves room for future extension into `let @` in global section if needed and conceptually it still makes sense to me, i.e. 'before  tracing,  create maps of these types explicitly'

**Comment by jordalgo** [comment]
>  in BEGIN only could also work

@fbs  I see what you're saying but part of me doesn't even want to have this restriction and just let semantic analyser detect if one of these "initializers" is being called multiple times on the same map. Then we don't force users into specifying a whole `BEGIN` probe to use a certain type of map (per the "terseness" principal of the language).

**Comment by jordalgo** [comment]
Summarizing some notes from the past office hours. It seems like folks are good with some form of the initializer syntax that @fbs and @amscanne suggested. As this doesn't require new syntax (mostly).

The current debate now is whether to allow this syntax inside any probe at any point e.g.
```
kprobe:vfs_read { @my_map = hashmap(100); }
```

OR limit it to the global scope e.g.
```
let @my_map = hashmap(100);
kprobe:vfs_read { ... }
```

Even though the global scope option may be slightly more complicated in terms of implementation, I'm personally in favor of this option (despite my most recent comment) as well as @danobi and potentially @fbs. As @danobi called out in the last meeting, allowing this syntax in some kind of control flow, when it's really not generating any instructions at that point in the code, would be confusing. The global scope option is slightly less terse but I feel like it's much more intuitive and readable. We also have already introduced the `let` syntax so we might as well use it for maps/globals as well as scratch/local variables.

**Comment by fbs** [comment]
Either seems fine to me. Global would have my preference for the reasons you have their but indeed needs the concept of a 'global section'

**Comment by jordalgo** [comment]
Initial POC for folks to play with. It adds map declarations in the global scope and it seems pretty self-contained: https://github.com/bpftrace/bpftrace/pull/3863

**Comment by jordalgo** [comment]
Initial POC has landed with an "unstable" flag. Going to close this for now - we can re-open if there end up being issues with the implementation or syntax.
